### PR TITLE
add support for `--extra-source-urls` to fetch sources from additional URLs

### PR DESCRIPTION
**Pull request by joeydumont**
This PR adds a --sources-url CLI option. It is a comma separated list of URLs that EasyBuild will fetch sources from. It replaces the hard-coded `EASYBUILD_SOURCES_URL`, but keeps it as a default value. The --sources-url-priority option is used to determine if these URLs are to be queried first, or last, i.e. as a fallback.

Unfortunately, this breaks 3 tests related to downloading the file, and I'm unsure of exactly why. It seems that `build_option('sources_url')` might return None when the function is run through the test harness, rather than getting populated with the default value. I've tried adding a `init_config(args=["--sources-url=https://sources.easybuild.io"])`, but that didn't work either.

If the feature is of interest, could you give me a hand in fixing the tests?

EDIT: After some discussion on Slack, I was able to fix the tests. I also renamed the options to `--extra-source-urls` and `--extra-source-urls-priority`. I changed the commit mesage to be imperative.

### CODE REVIEW

File: easybuild/tools/options.py [hunk 1]
```diff
@@ -394,6 +396,11 @@ def override_options(self):
                              None, 'store_true', False),
             'extra-modules': ("List of extra modules to load after setting up the build environment",
                               'strlist', 'extend', None),
+            "extra-source-urls": ("Specify different URLs to fetch sources from", "strlist", "store",
```
Comment (Flamefire) [inline easybuild/tools/options.py:399]:
```suggestion
            "extra-source-urls": ("Specify URLs to fetch sources from in addition to those in the easyconfig", "strlist", "store",
```

File: easybuild/tools/options.py [hunk 2]
```diff
@@ -394,6 +396,11 @@ def override_options(self):
                              None, 'store_true', False),
             'extra-modules': ("List of extra modules to load after setting up the build environment",
                               'strlist', 'extend', None),
+            "extra-source-urls": ("Specify URLs to fetch sources from in addition to those in the easyconfig",
+                                  "strlist", "store", DEFAULT_EXTRA_SOURCE_URLS),
+            "extra-source-urls-priority": ("Specify whether sources-url should be tried before or after the "
```
Comment (jfgrimm) [inline easybuild/tools/options.py:401]:
```suggestion
            "extra-source-urls-priority": ("Specify whether extra-source-urls should be tried before or after the "
```

Comment (valandil) [reply easybuild/tools/options.py]: Pushed that.

Comment (joeydumont) [reply easybuild/tools/options.py]: Pushed that.

**Comment by joeydumont** [comment]
I see there are now merge conflicts for this. Let me know when you're ready for me to rebase, or if you prefer to handle that.

**Comment by jfgrimm** [comment]
> I see there are now merge conflicts for this. Let me know when you're ready for me to rebase, or if you prefer to handle that.

go for it; if you have any issues/questions just ping me

**Comment by Flamefire** [comment]
Suggestion here:
`--sources-url` supports multiple URLs, doesn't it? How are they separated?

For EasyConfigs we have --robot-paths which accepts a list of paths like  /path1:/path2 and works nicely mit the default paths if you append or prepend a colon: :/path1:/path2 appends it to the default, /path1:/path2: prepends it while the original replaces it.
How about replacing the priority option by something like this to have a uniform interface here?

**Comment by joeydumont** [comment]
As suggested, I made sure that this used the same `add_flex` logic that `--robot-paths` uses. However, I think URLs should be separated by `|` rather than a colon (`:`) or a comma (,), as both are valid in URLs. I added `urllist` to the list of options in `ExtOption` to allow for this separator to be used.

File: easybuild/framework/easyblock.py [hunk 3]
```diff
@@ -897,8 +895,13 @@ def obtain_file(self, filename, extension=False, urls=None, download_filename=No
                     source_urls = []
                 source_urls.extend(self.cfg['source_urls'])
 
-                # add https://sources.easybuild.io as fallback source URL
-                source_urls.append(EASYBUILD_SOURCES_URL + '/' + os.path.join(name_letter, location))
+                # add extra-source-urls CLI as either a first check, or a fallback.
+                self.log.warning("[extra_source_urls] %s", build_option('extra_source_urls'))
```
Comment (Flamefire) [inline easybuild/framework/easyblock.py:899]: Why are those warnings? I'd suggest to only log if there are any at all, as (at most) `info` and the individual urls as `debug`. And maybe only log the final value, not each intermediate

File: easybuild/framework/easyblock.py [hunk 4]
```diff
@@ -897,8 +895,13 @@ def obtain_file(self, filename, extension=False, urls=None, download_filename=No
                     source_urls = []
                 source_urls.extend(self.cfg['source_urls'])
 
-                # add https://sources.easybuild.io as fallback source URL
-                source_urls.append(EASYBUILD_SOURCES_URL + '/' + os.path.join(name_letter, location))
+                # add extra-source-urls CLI as either a first check, or a fallback.
```
Comment (Flamefire) [inline easybuild/framework/easyblock.py:898]: The comment seems outdated. I don't see an either-or in the code below. Also CLI should rather be "option" or "command line option" as CLI is usually the whole set of (all) CL options

File: easybuild/framework/easyblock.py [hunk 5]
```diff
@@ -897,8 +895,13 @@ def obtain_file(self, filename, extension=False, urls=None, download_filename=No
...
+                self.log.warning("[extra_source_urls] %s", build_option('extra_source_urls'))
+                for url in build_option("extra_source_urls"):
+                    url += "/" + name_letter + "/" + location
+                    self.log.warning("[extra_source_urls] url is %s", url)
+                    source_urls.extend([url])
```
Comment (Flamefire) [inline easybuild/framework/easyblock.py:903]: It's only a single element:
```suggestion
                    source_urls.append(url)
```

File: easybuild/tools/options.py [hunk 6]
```diff
@@ -407,6 +409,8 @@ def override_options(self):
                              None, 'store_true', False),
             'extra-modules': ("List of extra modules to load after setting up the build environment",
                               'strlist', 'extend', None),
+            "extra-source-urls": ("Specify URLs to fetch sources from in addition to those in the easyconfig",
+                                  "urltuple", "add_flex", DEFAULT_EXTRA_SOURCE_URLS, {'metavar': 'URL[%sURL]' % '|'}),
```
Comment (Flamefire) [inline easybuild/tools/options.py:413]: I'd rather not use string interpolation for a constant. Just makes loading EB (even) slower:
```suggestion
                                  "urltuple", "add_flex", DEFAULT_EXTRA_SOURCE_URLS, {'metavar': 'URL[|URL]'}),
```

**Review by Flamefire — CHANGES_REQUESTED** [review]

File: easybuild/tools/options.py (hunk 6 above)
Comment (joeydumont) [reply easybuild/tools/options.py]: Good point, done.

File: easybuild/framework/easyblock.py (hunk 5 above)
Comment (joeydumont) [reply easybuild/framework/easyblock.py]: Done.

File: easybuild/framework/easyblock.py (hunk 3 above)
Comment (joeydumont) [reply easybuild/framework/easyblock.py]: Thanks, I added those in for me to make sure the values were making their way in the expected places. I removed these logs completely, the debug logs already print
```
== 2024-07-16 20:58:08,626 generaloption.py:1604 DEBUG generate_cmd_line adding extra-source-urls value ('https://sources.easybuild.io/',) default found. Not adding to args.
```

File: easybuild/framework/easyblock.py (hunk 4 above)
Comment (joeydumont) [reply easybuild/framework/easyblock.py]: It was, fixed.

File: easybuild/framework/easyblock.py [hunk 7]
```diff
@@ -897,8 +895,11 @@ def obtain_file(self, filename, extension=False, urls=None, download_filename=No
                     source_urls = []
                 source_urls.extend(self.cfg['source_urls'])
 
-                # add https://sources.easybuild.io as fallback source URL
-                source_urls.append(EASYBUILD_SOURCES_URL + '/' + os.path.join(name_letter, location))
+                # Insert --extra-source-urls command line option to the
+                # urls to try to download from.
```
Comment (Flamefire) [inline easybuild/framework/easyblock.py:899]: As mentioned earlier this is not only a CLI option, but can be from 2 other places. I suggest something like:

```suggestion
                # Add additional URLs as configured
```

File: easybuild/base/generaloption.py [hunk 8]
```diff
@@ -182,6 +187,7 @@ class ExtOption(CompleterOption):
           - strlist, strtuple : convert comma-separated string in a list resp. tuple of strings
           - pathlist, pathtuple : using os.pathsep, convert pathsep-separated string in a list resp. tuple of strings
               - the path separator is OS-dependent
+          - urllist, urltuple: using '|', convert urlsep separated string in a list resp. tuple of strings
```
Comment (Flamefire) [inline easybuild/base/generaloption.py:190]: Nit if you want to touch that: Although it is a copy of the above there is no definition of "urlsep". So maybe:

```suggestion
          - urllist, urltuple: convert string separated by '|', to a list resp. tuple of strings
```

File: easybuild/framework/easyblock.py [hunk 9]
```diff
@@ -897,8 +895,10 @@ def obtain_file(self, filename, extension=False, urls=None, download_filename=No
                     source_urls = []
                 source_urls.extend(self.cfg['source_urls'])
 
-                # add https://sources.easybuild.io as fallback source URL
-                source_urls.append(EASYBUILD_SOURCES_URL + '/' + os.path.join(name_letter, location))
+                # Add additional URLs as configured.
+                for url in build_option("extra_source_urls") or []:
```
Comment (Flamefire) [inline easybuild/framework/easyblock.py:899]: Looks like f4ec1825c740410f667342caa3403a4e37657c33 is not required after you fixed the issue with 1f5542e586ad3661c56d7394786627f6661a5bc5, i.e. "extra_source_urls" can never be None hence no need for the `or []`

I'd suggest to either revert that commit or rebase skipping it. Otherwise people might get confused later how it could be anything but a (possibly) empty list. But it can't even be an **empty** list, can it?

Or:
```suggestion
                for url in build_option("extra_source_urls"):
```

File: easybuild/tools/config.py [hunk 10]
```diff
@@ -120,6 +120,7 @@
 DEFAULT_PR_TARGET_ACCOUNT = 'easybuilders'
 DEFAULT_PREFIX = os.path.join(os.path.expanduser('~'), ".local", "easybuild")
 DEFAULT_REPOSITORY = 'FileRepository'
+DEFAULT_EXTRA_SOURCE_URLS = ('https://sources.easybuild.io/',)
```
Comment (Flamefire) [inline easybuild/tools/config.py:123]: The trailing slash lead to a duplication later (added in the loop using it, wasn't there before) on which might cause issues, see above: `url += "/" + name_letter + "/" + location`

```suggestion
DEFAULT_EXTRA_SOURCE_URLS = ('https://sources.easybuild.io',)
```

File: easybuild/tools/options.py [hunk 11]
```diff
@@ -69,10 +69,12 @@
...
 from easybuild.tools.config import DEFAULT_MINIMAL_BUILD_ENV, DEFAULT_MNS, DEFAULT_MODULE_SYNTAX, DEFAULT_MODULES_TOOL
 from easybuild.tools.config import DEFAULT_MODULECLASSES, DEFAULT_PATH_SUBDIRS, DEFAULT_PKG_RELEASE, DEFAULT_PKG_TOOL
-from easybuild.tools.config import DEFAULT_PKG_TYPE, DEFAULT_PNS, DEFAULT_PREFIX, DEFAULT_PR_TARGET_ACCOUNT
+from easybuild.tools.config import DEFAULT_PKG_TYPE, DEFAULT_PNS, DEFAULT_PREFIX, DEFAULT_EXTRA_SOURCE_URLS
+from easybuild.tools.config import DEFAULT_PR_TARGET_ACCOUNT
 from easybuild.tools.config import DEFAULT_REPOSITORY, DEFAULT_WAIT_ON_LOCK_INTERVAL, DEFAULT_WAIT_ON_LOCK_LIMIT
 from easybuild.tools.config import DEFAULT_FILTER_RPATH_SANITY_LIBS
-from easybuild.tools.config import EBROOT_ENV_VAR_ACTIONS, ERROR, FORCE_DOWNLOAD_CHOICES, GENERAL_CLASS, IGNORE
+from easybuild.tools.config import EBROOT_ENV_VAR_ACTIONS, ERROR
+from easybuild.tools.config import FORCE_DOWNLOAD_CHOICES, GENERAL_CLASS, IGNORE
```
Comment (Flamefire) [inline easybuild/tools/options.py:77]: Why this seemingly unrelated change?
```suggestion
from easybuild.tools.config import EBROOT_ENV_VAR_ACTIONS, ERROR, FORCE_DOWNLOAD_CHOICES, GENERAL_CLASS, IGNORE
```

File: easybuild/base/generaloption.py [hunk 12]
```diff
@@ -103,6 +103,11 @@ def what_str_list_tuple(name):
         sep = os.pathsep
         helpsep = 'pathsep'
```
Comment (Flamefire) [inline easybuild/base/generaloption.py:105]: Can you remove this empty line please? And as we have 3 cases now, I think it would be great to put the comma part above into an `else` clause below this (again without empty line between previous "helpset" and "else")

**Review by Flamefire — CHANGES_REQUESTED** [review]
A final round of comments. All minor things, so it's just polishing this up. :+1:

File: easybuild/framework/easyblock.py (hunk 9 above)
Comment (joeydumont) [reply easybuild/framework/easyblock.py:899]: My understanding is that it cannot be an empty list now it has a properly listed default value in `BUILD_OPTIONS_CMDLINE`, indeed. I'll remove it.

You mentioned rebasing. Would you prefer I squash everything to one commit here? Or do you want to keep the review history?

File: easybuild/tools/options.py (hunk 11 above)
Comment (joeydumont) [reply easybuild/tools/options.py:77]: I had additional imports in the previous implementation, and they were subsequently removed. I've cleaned that up.

File: easybuild/tools/options.py [hunk 13]
```diff
@@ -69,9 +69,9 @@
 from easybuild.tools.config import DEFAULT_JOB_EB_CMD, DEFAULT_LOGFILE_FORMAT, DEFAULT_MAX_FAIL_RATIO_PERMS
 from easybuild.tools.config import DEFAULT_MINIMAL_BUILD_ENV, DEFAULT_MNS, DEFAULT_MODULE_SYNTAX, DEFAULT_MODULES_TOOL
 from easybuild.tools.config import DEFAULT_MODULECLASSES, DEFAULT_PATH_SUBDIRS, DEFAULT_PKG_RELEASE, DEFAULT_PKG_TOOL
-from easybuild.tools.config import DEFAULT_PKG_TYPE, DEFAULT_PNS, DEFAULT_PREFIX, DEFAULT_PR_TARGET_ACCOUNT
+from easybuild.tools.config import DEFAULT_PKG_TYPE, DEFAULT_PNS, DEFAULT_PREFIX, DEFAULT_EXTRA_SOURCE_URLS
 from easybuild.tools.config import DEFAULT_REPOSITORY, DEFAULT_WAIT_ON_LOCK_INTERVAL, DEFAULT_WAIT_ON_LOCK_LIMIT
-from easybuild.tools.config import DEFAULT_FILTER_RPATH_SANITY_LIBS
+from easybuild.tools.config import DEFAULT_FILTER_RPATH_SANITY_LIBS, DEFAULT_PR_TARGET_ACCOUNT
```
Comment (Flamefire) [inline easybuild/tools/options.py:74]: Just FYI: We try to keep the imports sorted. So the ordering here would be the other way round. But not really important

File: easybuild/framework/easyblock.py (hunk 9 above)
Comment (Flamefire) [reply easybuild/framework/easyblock.py]: That is correct, yes

I'm not allowed to merge anyway so I can't really decide it. But my person preference for this is to squash. Most commits are just stylistic anyway, not some redesign or something. Only thing would be https://github.com/easybuilders/easybuild-framework/pull/4079/commits/1f5542e586ad3661c56d7394786627f6661a5bc5 to keep a mention why it is put there but even that is "obvious" enough. Hence a single commit is fine for me.

But I know that others like to keep all history or just don't want to do the extra work of squashing.

**Review by Flamefire — APPROVED** [review]

File: easybuild/tools/config.py [hunk 14]
```diff
@@ -120,6 +120,7 @@
 DEFAULT_PR_TARGET_ACCOUNT = 'easybuilders'
 DEFAULT_PREFIX = os.path.join(os.path.expanduser('~'), ".local", "easybuild")
 DEFAULT_REPOSITORY = 'FileRepository'
+DEFAULT_EXTRA_SOURCE_URLS = ('https://sources.easybuild.io',)
```
Comment (boegel) [inline easybuild/tools/config.py:123]:
```suggestion
EASYBUILD_SOURCES_URL = 'https://sources.easybuild.io'
DEFAULT_EXTRA_SOURCE_URLS = (EASYBUILD_SOURCES_URL,)

File: easybuild/framework/easyblock.py [hunk 15]
```diff
@@ -106,8 +106,6 @@
 from easybuild.tools.version import this_is_easybuild, VERBOSE_VERSION, VERSION
 
 
-EASYBUILD_SOURCES_URL = 'https://sources.easybuild.io'
```
[the remark this answers is not in the document]
Comment (joeydumont) [reply easybuild/framework/easyblock.py:109]: We can indeed define it in `easybuild.tools.config`, but then importing it in `easybuild.framework.easyblock` and not using it makes houndbot complain. I defined it separately in two locations to fix houndbot, but it's not optimal.

[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/framework/easyblock.py:109]: @joeydumont Hmm, I would be surprised if the bot complains about a constant not being used in `easybuild.tools.config`, but if it does, you can add a comment to skip the style check for that line:

```python
EASYBUILD_SOURCES_URL = 'https://sources.easybuild.io'  # noqa

[the remark this answers is not in the document]
Comment (joeydumont) [reply easybuild/framework/easyblock.py:109]: I meant in `easybuild.framework.easyblock`, if we had
```
from easybuild.tools.config import EASYBUILD_SOURCES_URL
```
so that it doesn't break the API. That's why I've done now, but added `# noqa`. 

In short, now `EASYBUILD_SOURCES_URL` is defined in `easybuild.tools.config`, where it is used, and imported in `easybuild.framework.easyblock`, where it used to be defined, so it can still be imported as `from easybuild.framework.easyblock`, but defined in a single place. The `# noqa` is required for HoundBot not to complain about the unused variable in `easybuild.framework.easyblock`.

[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/framework/easyblock.py:109]: :+1:

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
Good to go, thanks a lot for your patience with this one @joeydumont, I'm happy it's finally getting merged :)

**Comment by joeydumont** [comment]
> Good to go, thanks a lot for your patience with this one @joeydumont, I'm happy it's finally getting merged :)

And thank you! I appreciate the feedback!
